src/main/scala/io/iohk/ethereum/jsonrpc/FilterManager.scala

```diff
@@ -0,0 +1,271 @@
+          case None => logsSoFar
+        }
+      }
+    }
+
+    val fromBlockNumber =
+      startingBlockNumber.getOrElse(resolveBlockNumber(filter.fromBlock.getOrElse(BlockParam.Latest), appStateStorage))
+
+    val toBlockNumber =
+      resolveBlockNumber(filter.toBlock.getOrElse(BlockParam.Latest), appStateStorage)
+
+    recur(fromBlockNumber, toBlockNumber, Nil)
+  }
+
+  private def getFilterChanges(id: BigInt): Unit = {
+    val bestBlockNumber = appStateStorage.getBestBlockNumber()
+    val lastCheckBlock = lastCheckBlocks.getOrElse(id, bestBlockNumber)
+
+    val filterOpt = filters.get(id)
+    filterOpt.foreach { _ => lastCheckBlocks += (id -> bestBlockNumber) }
+    resetTimeout(id)
+
+    filterOpt match {
+      case Some(logFilter: LogFilter) =>
+        sender() ! LogFilterChanges(getLogs(logFilter, Some(lastCheckBlock + 1)))
+
+      case Some(_: BlockFilter) =>
+        sender() ! BlockFilterChanges(getBlockHashesAfter(lastCheckBlock).takeRight(maxBlockHashesChanges))
+
+      case Some(_: PendingTransactionFilter) =>
+        getPendingTransactions().map { pendingTransactions => // TODO: should this return only pending transactions added since last poll?
+          PendingTransactionFilterChanges(pendingTransactions.map(_.hash))
+        }.pipeTo(sender())
+
+      case None =>
+        sender() ! LogFilterChanges(Nil)
+    }
+  }
+
+  private def getLogsFromBlock(filter: LogFilter, blockHeader: BlockHeader, receipts: Seq[Receipt]): Seq[Log] = {
+    val bytesToCheckInBloomFilter = filter.address.map(a => Seq(a.bytes)).getOrElse(Nil) ++ filter.topics.flatten
+
+    receipts.zipWithIndex.foldLeft(Seq[Log]()) { case (logsSoFar, (receipt, txIndex)) =>
+      if (BloomFilter.containsAnyOf(receipt.logsBloomFilter, bytesToCheckInBloomFilter)) {
+        logsSoFar ++ receipt.logs.zipWithIndex
+        .filter { case (log, _) => filter.address.forall(_ == log.loggerAddress) && topicsMatch(log.logTopics, filter.topics) }
+        .map { case (log, logIndex) =>
+          val blockBody = blockchain.getBlockBodyByHash(blockHeader.hash).get
+          val tx = blockBody.transactionList(txIndex)
+          Log(
+            logIndex = logIndex,
+            transactionIndex = Some(txIndex), // TODO: None is this is a pending block (?)
+            transactionHash = Some(tx.hash), // TODO: None is this is a pending block (?)
+            blockHash = blockHeader.hash,
+            blockNumber = blockHeader.number,
+            address = log.loggerAddress,
+            data = log.data,
+            topics = log.logTopics)
+        }
+      } else logsSoFar
+    }
+  }
+
+  private def topicsMatch(logTopics: Seq[ByteString], filterTopics: Seq[Seq[ByteString]]): Boolean = {
+    (filterTopics zip logTopics).forall { case (filter, log) => filter.isEmpty || filter.contains(log) }
+  }
+
+  private def getBlockHashesAfter(blockNumber: BigInt): Seq[ByteString] = {
+    val bestBlock = appStateStorage.getBestBlockNumber()
```

Comment:
As this is being queried here https://github.com/input-output-hk/etc-client/pull/218/files#diff-b5ee4e4c8bcd462269de630eb5466a06R127 maybe we can send it as paramenter instead of getting that value again